packages/connect/src/base/listener.ts

```diff
@@ -215,30 +218,43 @@ class Listener extends EventEmitter implements InterfaceListener {
         .join(`\n\t`)}`
     )
 
-    // Rebuild later
-    setImmediate(this.updatePublicNodes.bind(this))
+    // Only rebuild list of relay nodes if we were using the
+    // offline node
+    if (inUse) {
+      // Rebuild later
+      setImmediate(this.updatePublicNodes.bind(this))
+    }
   }
 
   /**
    * Updates the list of exposed entry nodes.
    * Called at startup and once an entry node is considered offline.
    */
   protected async updatePublicNodes(): Promise<void> {
+    const knownNodes = new Set<string>(this.publicNodes.map((entry: NodeEntry) => entry.id.toB58String()))
     const nodesToCheck: PeerStoreType[] = []
 
     for (const uncheckedNode of this.uncheckedNodes) {
       if (uncheckedNode.id.equals(this.peerId)) {
-        return
+        continue
       }
 
-      for (const publicNode of this.publicNodes) {
-        if (publicNode.id.equals(uncheckedNode.id)) {
-          // Peer is already a public node, so nothing to do
-          return
+      const usableAddresses: Multiaddr[] = uncheckedNode.multiaddrs.filter(isUsableRelay)
+
+      if (knownNodes.has(uncheckedNode.id.toB58String())) {
+        const index = this.publicNodes.findIndex((entry) => entry.id.equals(uncheckedNode.id))
+
+        if (index < 0) {
```

Comment:
This should not happen due to the previous conditions, right?

Reply:
Don't think so. `onNewRelay` populates the `uncheckedNodes` array without checking whether we know the relay node or not. Once we rebuild the list, we consume all nodes on the list and rebuild the list of entry nodes. The result is an empty `uncheckNodes` array.

Reply:
Adding nodes more than once seems to happen suprisingly often, namely when our cloud get restarted, e.g. after a commit to a release branch